tools/clang/lib/SPIRV/SpirvEvalInfo.h

```diff
@@ -0,0 +1,103 @@
+//===------- SpirvEvalInfo.h - SPIR-V Evaluation Info -----------*- C++ -*-===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//===----------------------------------------------------------------------===//
+//
+//  This file defines a struct containing SPIR-V information from evaluating
+//  Clang AST tree.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_LIB_SPIRV_SPIRVEVALINFO_H
+#define LLVM_CLANG_LIB_SPIRV_SPIRVEVALINFO_H
+
+#include "spirv/1.0/spirv.hpp11"
+
+namespace clang {
+namespace spirv {
+
+/// Memory layout rules
+enum class LayoutRule {
+  Void,
+  GLSLStd140,
+  GLSLStd430,
+};
+
+/// Struct contains SPIR-V information from evaluating a Clang AST tree.
+///
+/// We need to report more information than just the <result-id> for SPIR-V:
+/// * Constness: enables generating SPIR-V constant instructions
+/// * Storage class: for getting correct (pointer) types
+/// * Layout rule: for getting correct (pointer) types
+/// * Relaxed precision: for emitting PrelaxedPrecision decoration
+///
+/// For most cases, the evaluation result of a Clang AST tree will just be
+/// a SPIR-V <result-id> that is not a constant, with storage class Function,
+/// layout rule Void, and not relaxed precision.
+///
+/// The evaluation result of a Clang Expr can be considered as SPIR-V constant
+/// as long as all its operands are SPIR-V constants and we are creating a
+/// composite out of them.
+///
+/// We derive storage class or layout rule from a base variable. Access chains
+/// and load/store/insertion/extraction of the base variable retains the same
+/// storage class or layout rule. Other operations will likely turn the storage
+/// class into Function and layout rule into Void.
+///
+/// Relaxed precision decorated on a variable propagates to all operations
+/// involving that variable.
```

Comment:
I have a question:
For the following:
```
%result = OpIAdd %int %a %b
```
Should `%result` have RelaxedPrecision decoration if

a) One of %a **or** %b are RelaxedPrecision

b) Both %a **and** %b are RelaxedPrecision

Reply:
I think its a). glslang confirms that. But I need to find spec about this.

Reply:
SGTM